client/build_test.go

```diff
@@ -175,3 +184,367 @@ func testUnknownBuildID(t *testing.T, sb integration.Sandbox) {
+			Args:   []string{"cat", "/tmp/test"},
+			Cwd:    "/",
+			Stdout: &nopCloser{output},
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		err = pid4.Wait()
+		if err != nil {
+			return nil, err
+		}
+
+		return &client.Result{}, nil
+	}
+
+	_, err = c.Build(ctx, SolveOpt{}, product, b, nil)
+	require.NoError(t, err)
+	require.Equal(t, "testing", output.String())
+
+	checkAllReleasable(t, c, sb, true)
+}
+
+// testClientGatewayContainerPID1Fail is testing clean shutdown and release
+// of resources when the primary pid1 exits with non-zero exit status
+func testClientGatewayContainerPID1Fail(t *testing.T, sb integration.Sandbox) {
+	requiresLinux(t)
+
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	product := "buildkit_test"
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal state")
+		}
+
+		r, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to solve")
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{
+			Mounts: []client.Mount{{
+				Dest:      "/",
+				MountType: pb.MountType_BIND,
+				Ref:       r.Ref,
+			}},
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		pid1, err := ctr.Start(ctx, client.StartRequest{
+			Args: []string{"false"},
+			Cwd:  "/",
+		})
+		if err != nil {
+			ctr.Release(ctx)
+			return nil, err
+		}
+
+		defer ctr.Release(ctx)
+		err = pid1.Wait()
+
+		var exitError *errdefs.ExitError
+		require.True(t, errors.As(err, &exitError))
+		require.Equal(t, uint32(1), exitError.ExitCode)
+
+		return nil, err
+	}
+
+	_, err = c.Build(ctx, SolveOpt{}, product, b, nil)
+	require.Error(t, err)
+
+	checkAllReleasable(t, c, sb, true)
+}
+
+// testClientGatewayContainerPID1Exit is testing that all process started
+// via `Exec` are shutdown when the primary pid1 process exits
+func testClientGatewayContainerPID1Exit(t *testing.T, sb integration.Sandbox) {
+	if sb.Rootless() {
+		// TODO fix this
+		// We get `panic: cannot statfs cgroup root` when running this test
+		// with runc-rootless
+		t.Skip("Skipping runc-rootless for cgroup error")
+	}
+	requiresLinux(t)
+
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	product := "buildkit_test"
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal state")
+		}
+
+		r, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to solve")
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{
+			Mounts: []client.Mount{{
```

Comment:
Could we have a test that uses more mounts. And maybe some custom mounts.

Reply:
I added a test to attach multiple mounts:
https://github.com/moby/buildkit/pull/1627/commits/516b9ec8445dadb83d89145295ff35d677d1ba8a

Not sure how to deal with tmpfs or cache mounts, we had discussed in the past moving the "ownership" of these mounts to the gateway so that we could get a reference to them for debugging on error.  For now we can a) ignore the tmpfs and cache mounts, b) reproduce the tmpfs, cache logic from libsolver ExecOp into the gateway.  We could also pause this PR to look into moving the ownership of the tmpfs, cache mounts to the gateway so we could reuse the Ref logic.  Thoughts?

Reply:
I think we should at least support all the mount types, even if they are not reusable across multiple exec atm. This would make sure we are compliant with `ExecState`. But it can be in the next PR if you prefer.

Reply:
Will try to update the PR with other mount types.

Reply:
for `ssh` and `secret` mounts, it seems the gateway will need access to a `session.Manager` and `session.Group`. Any ideas on where `gateway.NewContainer` would get access to these resources?

Reply:
Look at how this is done for `ResolveImageConfig` https://github.com/moby/buildkit/blob/master/solver/llbsolver/bridge.go#L284 . Logic should be the same I think.

Reply:
Gateway forwarder also seems to have access to session ID so that is another approach that might be simpler for the current layout of the packages.

Reply:
I have updated to allow for all mount types, it is a big commit:
https://github.com/moby/buildkit/pull/1627/commits/3cc8aa06498fb9a5e8ce6afd5e389d1b62e26cab
Most of the changes are moving the mounting logic from ExecOp into a new package so it can be reused for gateway exec, I can revert this commit if you think it is a terrible idea.  

I also had to add a `SessionManager() *session.Manager` to the the Worker interface, at least I could not figure out another way to get access to to it in the code and this seemed like the most straight forward way.

Reply:
I updated tests to include a secret mount and an ssh mount, so that covers all of them.
https://github.com/moby/buildkit/pull/1627/commits/5fcf6dd4267a7f552fc6a82a0affdc947f7319e2